std/array.d

```diff
@@ -3339,6 +3340,69 @@ if (isDynamicArray!A)
     app.put(r[]);
 }
 
+unittest // Issue 13300
+{ 
+    // Issue 13300: appender must work with non pure postblit
+    import std.array : array;
+
+    struct Simple13300
+    {
+        @disable this(); 
+        this(this) { import std.stdio : stdout; stdout; }  // Not pure!
+
+        private:
+        this(int tmp) { }
+    }
+
+    struct Range13300
+    {
+        @property front() { return Simple13300(0); }
+        void popFront() { count++; }
+        @property empty() { return count > 3; }
+        size_t count;
+    }
+
+    Range13300 r;
+    assert(r.array.length == 4);
+}
+
+@safe pure unittest // Issue 13300.
+{
+    // Issue 13300: appender must work with pure code as well
+    import std.array : array;
+   
+    struct Simple13300(bool disableDefaultCtor)
+    {
+        pure:
+        static if(disableDefaultCtor) @disable this();
+        this(this) {  }
+
+        private:
+        this(int tmp) { }
+    }
+
+    struct Range13300(bool disableDefaultCtor)
+    {
+        pure:
+
+        @property front() { return Simple13300!disableDefaultCtor(0); }
+        void popFront() { count++; }
+        @property empty() { return count > 3; }
+        size_t count;
+    }
+
+    // Test ensureAddable workaround for disable @this().
+    {
+        Range13300!true r;
+        assert(r.array.length == 4);
+    }
+
+    {
+        Range13300!false r;
+        assert(r.array.length == 4);
+    }
```

Comment:
Similarly here, why are you doing all of this stuff with `@disable this();`? Isn't this just supposed to be testing that the code is still `pure` if the postblit is `pure`? It seems to me that this code should be simplified so that that's what it's testing and use `only` to shorten it. All of this could just be a few lines of code. And of course, it needs to test that it's `@safe` and `nothrow` with those attributes removed as mentioned above.

Reply:
On ensureAddable() there's a workaround for struct with default ctor disabled. That's where the bug pop out. So I check both with and without default ctor disabled to be sure it works in both cases.